### PR TITLE
Feature: Added support for No sql database support and database migration support

### PR DESCRIPTION
**Pull request by rudreshgp**
Hi Boriszn,

 Good Day!
Added support for No SQL database (Tested locally for Postgres) and support for Migrations. 

So based on the appSettings.json either NoSQL or SQLServer database is used.

Thanks,
Rudresh

### CODE REVIEW

**Comment by Boriszn** [comment]
Hi @rudreshgp 

Thank you a lot, looks quite good.  Could you please add detail description to setup no NoSql database locally etc, in Readme file. (Perhaps, with checklist with howto instruction) ? 

Many Thanks.

File: src/DeviceManager.Api/Configuration/EntityFrameworkConfiguration.cs [hunk 1]
```diff
@@ -17,13 +21,44 @@ public static class EntityFrameworkConfiguration
...
-            string connectionString = configuration.GetConnectionString("DefaultConnection");
+            string connectionString = configuration.GetConnectionString(Constants.DefaultConnection);
+
+            // Database connection settings
+            var connectionOptions = services.BuildServiceProvider().GetRequiredService<IOptions<ConnectionSettings>>();
 
             // Entity framework configuration
-            services.AddDbContext<DeviceContext>(options =>
-                options.UseSqlServer(connectionString));
+            services.AddEntityFrameworkNpgsql().AddDbContext<DeviceContext>(options =>
```
Comment (Boriszn) [inline src/DeviceManager.Api/Configuration/EntityFrameworkConfiguration.cs]: Very nice. Thanks. Just wandering, why you choose PostgreSQL ? There are other powerful nosql like MongoDb, CouchDB etc.

File: src/DeviceManager.Api/Configuration/ConfigurationOptions.cs [hunk 2]
```diff
@@ -16,7 +17,7 @@ public static class ConfigurationOptions
         /// <param name="configuration">The configuration.</param>
         public static void ConfigureService(IServiceCollection services, IConfigurationRoot configuration)
         {
-            services.Configure<ConnectionSettings>(configuration.GetSection("ConnectionStrings"));
+            services.Configure<ConnectionSettings>(configuration.GetSection(Constants.ConnectionStrings));
```
Comment (Boriszn) [inline src/DeviceManager.Api/Configuration/ConfigurationOptions.cs]: Thanks for this

File: src/DeviceManager.Api/Data/Management/ContextFactory.cs [hunk 3]
```diff
@@ -71,22 +73,51 @@ private DbContextOptionsBuilder<DeviceContext> ChangeDatabaseNameInConnectionStr
...
             return contextOptionsBuilder;
         }
 
+        /// <summary>
+        /// Get Connection string builder instance based on the database type
+        /// </summary>
+        /// <returns></returns>
+        private DbConnectionStringBuilder GetConnectionBuilder()
+        {
+            if (connectionOptions.Value.UseNoSql)
```
Comment (Boriszn) [inline src/DeviceManager.Api/Data/Management/ContextFactory.cs]: In my opinion, here, would be better to use ternary operator `?:` like  `(condition ? first_expression : second_expression; )` What do you think ?

File: src/DeviceManager.Api/Configuration/EntityFrameworkConfiguration.cs (hunk 1 above)
Comment (rudreshgp) [reply src/DeviceManager.Api/Configuration/EntityFrameworkConfiguration.cs]: I will test with other no sql databases also. Then you can accept the pull request.

File: src/DeviceManager.Api/Data/Management/ContextFactory.cs (hunk 3 above)
Comment (rudreshgp) [reply src/DeviceManager.Api/Data/Management/ContextFactory.cs]: Yeah, I will fix all the changes you pointed out :-) 
By the way when are you planning to release next version of the framework??

Comment (Boriszn) [reply src/DeviceManager.Api/Data/Management/ContextFactory.cs]: Hey, I would like to do it after I merge this PR ;)

**Comment by rudreshgp** [comment]
Hi Borizn,
I am planning to move database connection string to individual classes and based on the name of the database the class is injected and settings (Use Postgres SQL, use SQL server) are added. Also, I am adding async await support for the API.

**Comment by Boriszn** [comment]
Hey. I didn't get it why put connection string to classes if it automatically substituted in the ContextFactory (and can reached from DataBaseManager) ? Could you please provide some examples, probably ? You can post it here, like Code Snippets. Or we can discuss it here: https://gitter.im/DeviceManager-Api/Architecture ?

**Comment by Boriszn** [comment]
One more thing, could you please do me a favor and add tickets which would like to work or already working, here in Kanban Board Project: https://github.com/Boriszn/DeviceManager.Api/projects/1 

What do you think ?

**Comment by rudreshgp** [comment]
"[RD-5] Add logging to files/console" and "[RD-7] Update Swagger (swashbuckle .net core)" --> Implemented in the pull request 15
[RD-11] Add entity framework seed -->  Implemented in the current pull request (18)

**Comment by Boriszn** [comment]
Very nice, thanks a lot. So I will add tickets for you to add MongoDb, CouchDb providers. About Docker I've already started implemented it You will see first results soon. But I will need assistance here I think. I will let you know ;) Is it ok for you ? (And meantime review your PR)

File: README.md [hunk 4]
```diff
@@ -19,6 +19,12 @@ Todo list, accomplished tasks, can be [found Here.](https://github.com/Boriszn/D
 5. (Optional) Run API integration tests
 6. Build / Run.
 
+## Database Connection
+`DatabaseType` field is used to specify the database type the application should connect.
+Currently, the framework contains connection information for `SqlExpress` and `Postgres`. The value of `DatabaseType` should come from `DatabaseType` enum (src\DeviceManager.Api\Configuration\Settings) and should match the class name inside (src\DeviceManager.Api\Configuration\DatabaseTypes) and implement `IDatabaseType` interface.
```
[the remark this answers is not in the document]
Comment (Boriszn) [reply README.md]: I think, fine for now ;)

**Pull request merged**